lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js

```diff
@@ -182,4 +188,59 @@
 
         return -1;
     });
+
+    platform.registerFunction(FunctionsEnum.ArrayFilter, function (callbackfn, thisArg) {
+        // ECMAScript 2017 #sec-array.prototype.filter
+        "use strict";
+
+        if (this === null || this === undefined) {
+            __chakraLibrary.raiseThis_NullOrUndefined("Array.prototype.filter");
+        }
+
+        let o;
+        let len
+        if (__chakraLibrary.isArray(this)) {
+            o = this;
+            len = o.length;
+        } else {
+            o = __chakraLibrary.Object(this);
+            len = __chakraLibrary.GetLength(o);
+        }
+        
+        if (typeof callbackfn != "function") {
+            __chakraLibrary.raiseFunctionArgument_NeedFunction("Array.prototype.filter");
+        }
+
+        let a = __chakraLibrary.arraySpeciesCreate(o, 0);
+        let k = 0;
+        let to = 0;
+
+        // Binding: If thisArg is provided, we need to do the equivalent of callbackfn.call(thisArg, ...), but we cannot use .call because Function.prototype.call can be overriden.
+        // Unfortinately, locally recording the original Function.protoype.call would need to call originalCall.call(callbackfn, thisArg, ...), which has the same .call problem. 
+        // Instead, in the case that the thisArg is provided, bind thisArg to callbackfn using a safe equivalent of callbackfn.bind(thisArg). In that case the callback will not be 
+        // inlinable, but alternative is to call into the runtime with callInstanceFunction on each call of the callback which is also not inlinable.
+        let boundCallback = callbackfn;
+        if (thisArg === undefined) {
+            boundCallback = callbackfn;
+        } else {
+            // safe equivalent of calling "callbackfn.bind(thisArg)"
+            boundCallback = __chakraLibrary.callInstanceFunc(__chakraLibrary.functionBind, callbackfn, thisArg);
+        }
+
+        while (k < len) {
+            if (k in o) {
+
+                let kValue = o[k];
+                if (boundCallback(kValue, k, o))
+                {
+                    a[to] = kValue;
```

Comment:
this is CreateDataPropertyOrThrow? will it perform the same as the spec wants it to?

Reply:
I believe this will, but @zenparsing can hopefully confirm that.

Reply:
@akroshg Good call, it's not quite the same. Assignment will utilize the prototype chain, whereas `filter` creates a data property directly on the output array.

```js
Object.defineProperty(Array.prototype, '0', {
  get() { return 980 },
  set(value) { console.log('called setter') },
});

// Does not log "called setter"
[1].filter(() => true);

// Logs "called setter"
let a = [];
a[0] = 1;
```

Reply:
I think the best course of action here is to use the CreateDataPropertyOrThrow in the case when species returns anything other than an Array.

---
In reply to: [187750568](https://github.com/Microsoft/ChakraCore/pull/5137#discussion_r187750568) [](ancestors = 187750568)